File: src/Token.py
from Namespace import Namespace
from src.Settings import LangData

class Position:
  def __init__(self, line, char):
    self.line, self.char = line, char
  def __str__(self):
    return f'[{self.line+1}:{self.char+1}]'

class Token:
  def __init__(self, text, pos, literal=False):
    self.text = text
    self.setPos(pos)
    self.updateType()

  def __str__(self):
    pos = str(self.pos)
    return f'<Token {self.type} \'{self.text}\'{" "*(12-len(self.text))} @{pos}>'

  def append(self, text):
    self.text += text
    self.updateType()

  def isEmpty(self):
    return self.text == ''

  def setPos(self, pos):
    self.pos = pos

  def updateType(self):
    updated = False
    langDataDict = {k:eval(f'LangData.{k}') for k in dir(LangData)
      if not k.startswith('_')}
    for key, val in langDataDict.items():
      if self.text.startswith(val):
        self.type = key
        updated = True
    if not updated:
      self.type = 'Text'

File: javascript/src/Token.js

class Position {
  constructor(l, c) {
    this.line = l;
    this.char = c;
  }
}

class Token {
  constructor(text, pos, literal=false) {
    this.text = text;
    this.setPos(pos);
    this.updateType();
  }
  append(text) {
    this.text += text;
    this.updateType();
  }
  isEmpty() {
    return !this.text.length;
  }
  setPos(pos) {
    this.pos = pos;
  }
  updateType() {
    let updated = false;
    for (let [key, val] of Object.entries(LangData)) {
      if (this.text.startsWith(val)) {
        this.type = key;
        updated = true;
      }
    }
    if (!updated) {
      this.type = 'Text';
    }
  }
}
